fw/http_sched_ratio.c

```diff
@@ -1035,32 +1045,52 @@ tfw_sched_ratio_del_grp(TfwSrvGroup *sg)
 	call_rcu(&ratio->rcu, tfw_sched_ratio_cleanup_rcu_cb);
 }
 
+static void
+tfw_sched_ratio_put_conn_data(struct rcu_head *rcu)
+{
+	TfwRatioSrvConnList *cl = container_of(rcu, TfwRatioSrvConnList, rcu);
+	kfree(cl);
+}
+
+/**
+ * Add anew or update descriptor for server.
+ *
+ * Can be called from user context on (re-)configuration or from SoftIRQ context
+ * as well on websocket connection establishment to renew actual connection list.
+ */
 static int
 tfw_sched_ratio_srvdesc_setup_srv(TfwServer *srv, TfwRatioSrvDesc *srvdesc)
 {
 	size_t size, ci = 0;
-	TfwSrvConn **conn, *srv_conn;
+	TfwSrvConn *srv_conn;
+	TfwRatioSrvConnList *cl_copy;
+	TfwRatioSrvConnList *cl = rcu_dereference_bh_check(srvdesc->cl, 1);
 
-	size = sizeof(TfwSrvConn *) * srv->conn_n;
-	if (!(srvdesc->conn = kzalloc(size, GFP_KERNEL)))
+	size = sizeof(TfwRatioSrvConnList) + sizeof(TfwSrvConn *) * srv->conn_n;
+	if (!(cl_copy = kzalloc(size, in_task() ? GFP_KERNEL : GFP_ATOMIC)))
 		return -ENOMEM;
 
-	conn = srvdesc->conn;
 	list_for_each_entry(srv_conn, &srv->conn_list, list) {
+		if (tfw_srv_conn_unscheduled(srv_conn))
+			continue;
 		if (unlikely(ci++ == srv->conn_n))
 			goto err;
-		*conn++ = srv_conn;
+
+		cl_copy->conns[ci-1] = srv_conn;
 	}
 	if (unlikely(ci != srv->conn_n))
 		goto err;
```

Comment:
It seems I was wrong with the proposal of the functions unification in both the schedulers.

Previously we called `tfw_sock_srv_new_conn()`, which adds a new connection to the connection list, only from process context (since we have constant number of server connections and just failover TCP connections inside them). Moreover,  `tfw_peer_add_conn()` uses a spin lock to add a connection. I believe we have a mutex on concurrent configuraion reloads, so we rely on the fact that `tfw_sched_ratio_add_srv()` always works in single context.

Now we can call this function concurrently with other websocket creation and with the configuration reload, so we definitely have races in all the places in the function. We can back to the separate functions and use the spin lock in the new function, but we'll also need to use the spin lock in all other functions iterating the connections list, which introduces lock contention.

What you're doing in the patch is almost solving https://github.com/tempesta-tech/tempesta/issues/710 , which is a hard task on it's own to upgrade the schedulers data structures without locks.

Initially with #755 I assumed that you just "steal" `sk` from `TfwSrvConn` and initiate the existing connection failovering mechanism.